CredentialsManager/App.cs

```diff
@@ -0,0 +1,161 @@
+
+    private static bool _isExitInvoked;
+
+    public static void Run()
+    {
+      while (!_isExitInvoked)
+      {
+        try
+        {
+          Manager.Initialize();
+          PrintMenu();
+          GetUserInput();
+        }
+        catch (MyFileNotFoundException ex)
+        {
+          Console.WriteLine(ex.Message);
+          break;
+        }
+        catch (DuplicateUserCredentialsException ex)
+        {
+          Console.WriteLine(ex.Message);
+          Pause("Press ENTER to exit..");
+          break;
+        }
+        catch (UserNameIsTakenException ex)
+        {
+          Console.WriteLine(ex.Message);
+          Pause();
+        }
+      }
+    }
+
+    private static IList<MenuItem> GetMenuItems()
+    {
+      return new List<MenuItem>
+      {
+        new MenuItem { ItemId = 1, Label = "Login", Action = Login, ConsoleKey = ConsoleKey.D1 },
+        new MenuItem { ItemId = 2, Label = "Register", Action = Register, ConsoleKey = ConsoleKey.D2 }
+      };
+    }
+
+    private static Menu CreateMenu()
+    {
+      return new Menu
+      {
+        Header = "Credentials Manager",
+        MenuItems = GetMenuItems(),
+        ExitItem = new MenuItem { ItemId = 3, Label = "Exit", Action = Exit, ConsoleKey = ConsoleKey.D3 }
+      };
+    }
+
+    private static void ConsoleInit(string header)
+    {
+      Console.CursorVisible = false;
+      Console.Clear();
+      Console.WriteLine(header);
+      Console.WriteLine();
+    }
+
+    private static void PrintMenu()
+    {
+      ConsoleInit(Header);
+
+      Console.WriteLine(MainMenu.ToString());
+    }
+
+    private static void GetUserInput()
+    {
+      ConsoleKey pressedKey;
+      do
+      {
+        pressedKey = Console.ReadKey(true).Key;
+        if (!MainMenu.IsValidKey(pressedKey)) continue;
+
+        ConsoleInit(MainMenu[pressedKey].Label);
+        MainMenu[pressedKey].Action();
+        Pause();
+        break;
+      } while (pressedKey != MainMenu.ExitItem.ConsoleKey);
+    }
+
+    private static void Login()
+    {
+      var userName = PromptString("Username: ");
+      var userPassword = PromptString("Password: ", false);
+
+      var isLoginSuccessful = Manager.Login(new Credentials(userName, userPassword));
+      var statusMessage = isLoginSuccessful ? LoginMessage : FailedLoginMessage;
+
+      Console.WriteLine(statusMessage);
+    }
+
+    private static void Register()
+    {
+      var userName = PromptString("Username: ");
+      var userPassword = PromptString("Password: ", false);
+
+      Manager.Register(new Credentials(userName, userPassword));
+      Console.WriteLine(RegistrationSuccessfulMessage);
+    }
+
+    private static void Exit()
+    {
+      _isExitInvoked = true;
+    }
+
+    private static string PromptString(string message)
+    {
+      string userInput;
+      do
+      {
+        Console.Write(message);
+        userInput = Console.ReadLine() ?? string.Empty;
```

Comment:
`Console.ReadLine` will never return null during normal operation.  The only time it would return null is if the standard input stream is EOF.  In Windows this doesn't really happen unless you are launching a process as a subprocess of another parent process (in that case the input of the child process could be controlled directly by the parent process).

In Linux this happens when we redirect with the pipe `|` operator (technically you can also redirect on Windows, especially if you're a powershell junky).

Anyways, I bring this all up because...if `Console.ReadLine` returns `null` then `Console.ReadLine` is ALWAYS going to return `null` from here on out.  Consider what that would mean for your program here.

Reply:
Thank you for the detailed explanation. This one was not intentional, I think I accepted a suggestion from the IDE (Rider) by mistake because after I deleted the coalesce operator it's suggesting to add it like you saw it here.

Reply:
I will  update it in commit [57ba93c](https://github.com/csinn/CSharp-From-Zero-To-Hero-v2/pull/43/commits/57ba93c3d8d6eefbd10ab94be489528fef7569bf)

Reply:
I inspected the method I found that Console.ReadLine() returns `string?`, this is why I was getting the suggestion.